Automattic/newspack-popups
# feat: improved frequency options

### All Submissions:






### Changes proposed in this Pull Request:



~**Note:** will require a rebase after #869 is merged.~ ✅ 

Adds new frequency options for displaying prompts based on pageviews and configurable time period if the `NEWSPACK_EXPERIMENTAL_READER_ACTIVATION` is enabled. Includes the following changes:

* **Frequency:** this existing dropdown lets you select several frequency presets, similarly to before. A few changes:
  - Changed "Once" preset to "Once a month". Because prompt interaction data is purged monthly, there's now no difference between "once" (implicitly per session) and "once per month".
  - Added a "Once a week" preset, because why not?
  - Removed the restriction that disallowed overlay prompts from using the "Every pageview" preset. Now, overlays can be shown to the same reader more frequently than once per 24 hours, so there's no reason to block the use of this preset.
  - Added an "Every 4th page view, up to 5x per month" preset, which matches the desired default behavior for overlay prompts as specified by Blue Lena.
  - Added a "Custom" option to let advanced users set the following new options manually. Note that if one of the other frequency presets is selected, the preset will override whatever values might exist or not exist in the various frequency option meta fields. This is so that sites that are already running prompts with one of the existing presets won't experience any unexpected changes in prompt behavior.
* **Max number of displays:** This new option defines the maximum number of times the prompt can be displayed within the specified time period. The default is `0`, which means no maximum (status quo).
* **Start after pageview:** This new option lets users specify a minimum number of pageviews a reader must have before the prompt will be displayed. For example, if this option is set to `2`, the reader will only start seeing the prompt on the third pageview. The default is `0`, which means the prompt will be displayed immediately (status quo).
* **Pageviews between displays:** This new option lets users specify the number of pageviews that must be logged in between prompt displays. For example, if this option is set to `3`, once the prompt is displayed, the reader must view three pages before the prompt can be displayed again (on the fourth view after last prompt display). The default is `0`, meaning the prompt will be displayed every page (status quo).
* **Reset counter per:** This new option lets users specify the time period after which the "Max number of displays" counter is reset. For example, if the prompt is set to show a max of 3 times per week, readers will only see the prompt a maximum of three times in a given one-week period. The default is `month`, which is status quo because prompt interaction data is purged monthly.
  - **Note:** Because specific pageview data (such as timestamp for each view) is purged hourly for performance reasons, the "Start after pageview" and "Pageviews between displays" options are based off of the reader's lifetime pageview counter. Consider the following example: a prompt is set to "Start after pageview" `2` and "Reset counter" per `month`. A reader visits the site only once, and then again after 32 days. On that second pageview after 32 days, they will NOT see the prompt because that's only the reader's second pageview. If they then visit the site again within the same month, they will see the prompt.
  - Another option is to keep detailed pageview event data for each reader for 30 days so we can reset pageview-related behavior as well, but this may result in bloated events table data.

Closes #902.

### How to test the changes in this Pull Request:

1. On `master`, publish some prompts (overlay and inline) with "Once", "Once a day", and "Every pageview" frequency options.
2. With the `NEWSPACK_EXPERIMENTAL_READER_ACTIVATION` flag OFF or not present in `wp-config.php`, check out this branch and https://github.com/Automattic/newspack-plugin/pull/1779. Test the pre-existing prompts in a new incognito session and confirm that they continue to work as expected.
3. Edit or create a new overlay prompt. Note that the frequency options behave identically as in `master`, except that the "Once" frequency is renamed "Once a month".
4. Add `define( 'NEWSPACK_EXPERIMENTAL_READER_ACTIVATION', true );` to `wp-config.php` and refresh the prompt editor. Note the new and relabeled options in the "Frequency" dropdown here and in the Campaigns wizard dashboard as well.

<img width="287" alt="Screen Shot 2022-07-06 at 3 53 52 PM" src="https://user-images.githubusercontent.com/2230142/177650011-ee7ec59c-0ad1-4742-879c-1e1f9ef0501d.png">

5. Test each preset option in a new incognito session. Note that you won't be able to really test per day/week/month presets unless you edit the `prompt_seen` events in the events table to backdate them a day/week/month.
6. Select the "Custom" frequency option and test each new option. Again, you won't be able to really test reset option unless you edit the `prompt_seen` events in the events table to backdate them a day/week/month. Note that if you select a preset, then select "Custom", the detailed frequency options will reflect the behavior of the last-selected preset.

<img width="272" alt="Screen Shot 2022-07-06 at 3 56 15 PM" src="https://user-images.githubusercontent.com/2230142/177650314-998d8f31-fac8-4282-986c-ce354362cce6.png">

7. Confirm that all unit tests pass. 👇 

### Other information:
